// @strapi/design-system/dist/EmptyStateLayout/EmptyStateLayout.js
import { jsxs as m, jsx as e } from "react/jsx-runtime";
import "react";
import t from "prop-types";
import r from "styled-components";
import { Box as o } from "../Box/Box.js";
import { Typography as l } from "../Typography/Typography.js";
const c = r(o)`
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
`, h = r(o)`
  svg {
    height: ${88 / 16}rem;
  }
`, i = ({ icon: a, content: n, action: d, hasRadius: p, shadow: s }) => /* @__PURE__ */ m(c, { padding: 11, background: "neutral0", hasRadius: p, shadow: s, children: [
  /* @__PURE__ */ e(h, { paddingBottom: 6, "aria-hidden": !0, children: a }),
  /* @__PURE__ */ e(o, { paddingBottom: 4, children: /* @__PURE__ */ e(l, { variant: "delta", as: "p", textColor: "neutral600", children: n }) }),
  d
] });
i.defaultProps = {
  action: void 0,
  hasRadius: !0,
  shadow: "tableShadow"
};
i.propTypes = {
  action: t.node,
  content: t.string.isRequired,
  hasRadius: t.bool,
  icon: t.node.isRequired,
  shadow: t.string
};
export {
  i as EmptyStateLayout
};
